### PR TITLE
Locations support

### LINKED ISSUE #478: Locations concept in Yorc
**As:** Yorc dev team
**I want to:** switch from the *infrastructure* concept to a *location* concept to Yorc
**So that:** This will make Yorc more flexible

**AC1:** A location allows to connect to an infrastructure
**AC2:** Location configuration (like how to connect to the infrastructure) is stored into Consul to allow to store it in a unique place for all Yorc instances of a given cluster
**AC3:** A Yorc cluster should be able to address multiple locations of the same type
**AC4:** A4C locations and Yorc locations should somehow be paired 
**AC5:** Update resources collectors API to take this into account

Background:
Currently Yorc has an infrastructures concept. Exactly one infrastructure of each type could be handled by a Yorc cluster. Those infrastructures should be configured into a config file. This config file should be consistent across instances of Yorc in a given cluster.

In this US we should be able to define some locations in Yorc. A location holds configuration info. They should be stored in a place shared by all Yorc instances of a given cluster (typically Consul). A location should be uniquely identifiable. A yorc cluster should be able to address multiple locations of the same type.

Topologies should somehow identify targeted locations. (A4C location and Yorc location should somehow be paired)

### PR DESCRIPTION
**Pull request by laurentganne**
# Pull Request description

## Description of the change

Removed property `Infrastructures` from the Yorc `Configuration`.
The configuration references now an optional path to a locations file providing the configuration of locations.
This location file if provided is used to populate consul with locations configuration, if and only if no location was already stored in Consul.
So this locations file is used by the first Yorc to start, the first time it starts.

Then the code performing the infrastructure resources provisioning is checking a node template metadata property `location`
to get the name of the location where to create the resource.
If no location is provided in the node template property, then the first location of the infrastructure type corresponding to the node type is selected. 

### What I did

Main changes:

`config/config.go`
Removed property `Infrastructures` from the `Configuration` type.
Introduced a new property `LocationsFilePath` providing the path to a file describing locations to define if no locations were defined yet in Consul.

`locations/location_mgr.go`
New code managing locations to:
- initialize locations from a file, if no location was already defined in consul
- create/remove/updated locations
- get properties for the first location of a given type
- get properties for the location referenced in the metadata of a given node name on a given deployment

`server/server.go`
If a locations file path is defined in Yorc Configuration, calling `location_mgr` to initialze locations in Consul for this file.

 
`prov/*/executor.go`
`prov/*/execution.go`
`prov/*/monitoring_job.go`
`prov/*/generator.go`
Instead of retrieving infrastructure properties from the infrastructure defined in Yorc configuration,
retrieving location properties from the metadata property `location` associated to a given node, or if no such property is defined, using the first location of the given infrastructure type
 
`commands/bootstrap/bootstrap.go`
Added a flag -l/--location to provide a location name

` commands/bootstrap/inputs.go`
Defining a location instead of an infrastructure for all types, except from Hosts Pools, not yet done.

` commands/bootstrap/resources/topology/topology_infrastructure_*.tmpl`
Change in strcuturres: moved from infrastructures to locations

` commands/bootstrap/setup.go`
For the local yorc used to bootstrap, generating a locations file work/locations.yorc.yaml
Referencing the path to this file in local yorc config file work/config.yorc.yaml in property LocationsFilePath

`commands/bootstrap/topology.go`
Removed Infrastructures property by a Locations property

`commands/bootstrap/resources/topology/tosca_types.zip`
Integrated changes from https://github.com/ystia/forge/tree/feature/GH-106-locations making ansible playbooks used in bootstrap create a location configuration

`commands/server.go`
Removed code managing infrastructures

**hostspool location**
as hostspool location is already managed by Yorc, an adapter has been implemented to handle locations of hosts pools, ie to delegate hostspool location handling to hostspool manager.

Hostspool manager has been changed to handle many hostspools instead of a single one.
Hostspool CLI has been changed to pass location name with --location flag (-l) as by instance:
**yorc hp apply my-config-host.yaml -l myLocation**
A new CLI has been implemented to list all hostspool locations:
**yorc hp locations**

Same changes in the hostspool REST API with location name in the REST endpoint.
CLI and REST API documentatiosn have been updated.

### How to verify it

Bootstrap a Yorc configuration, using an input file describing a location configuration, for example for OpenStack :

```YAML
locations:
  - name: myLocation
    type: openstack
    properties:
      auth_url: http://1.2.3.4:5000/
      default_security_groups:
      - myGroup1
      - default
      password: mypasswd
      private_network_name: private-net1
      region: RegionOne
      tenant_name: mytenant
      user_name: myuser
```

## Applicable Issues

Closes #478

### CODE REVIEW

File: commands/bootstrap/resources/topology/topology_infrastructure_openstack.tmpl [hunk 1]
```diff
@@ -1,6 +1,7 @@
 {{ define "Infrastructure" }}
-      type: org.ystia.yorc.infrastructure.OpenStackConfig
-      properties: {{formatAsYAML .Infrastructures.openstack 8}}
+      type: org.ystia.yorc.location.OpenStackConfig
```
Comment (loicalbertin) [inline commands/bootstrap/resources/topology/topology_infrastructure_openstack.tmpl]: Why type is `org.ystia.yorc.location.OpenStackConfig` here while `org.ystia.yorc.infrastructure.GoogleConfig` config for gcp and others?

File: commands/bootstrap/resources/topology/topology_infrastructure_openstack.tmpl [hunk 2]
```diff
@@ -1,6 +1,7 @@
...
+      properties: {{formatAsYAML .Location.Properties 8}}
+        location_name: "{{.Location.Name}}"
```
Comment (loicalbertin) [inline commands/bootstrap/resources/topology/topology_infrastructure_openstack.tmpl]: Why `location_name` appear for openstack and not for others?

Comment (laurentganne) [reply commands/bootstrap/resources/topology/topology_infrastructure_openstack.tmpl]: yes it needs to appear on all

File: prov/kubernetes/executor.go [hunk 3]
```diff
@@ -55,14 +57,27 @@ func (e *defaultExecutor) ExecAsyncOperation(ctx context.Context, conf config.Co
 		return nil, 0, err
 	}
 
-	if isNilValue(e.clientset) {
-		e.clientset, err = initClientSet(conf)
-		if err != nil {
-			return nil, 0, err
-		}
+	var locationProps config.DynamicMap
+	locationMgr, err := locations.NewManager(conf)
```
Comment (loicalbertin) [inline prov/kubernetes/executor.go]: Is there a good reason for recreating the manager for each execution? Does it worth caching it?

File: locations/location_mgr.go [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+	hpAdapter adapter.LocationAdapter
+}
+
+// LocationsDefinition represents the structure of an initialization file defining locations
+type LocationsDefinition struct {
+	Locations []LocationConfiguration `yaml:"locations" mapstructure:"locations" json:"locations"`
+}
+
+// NewManager creates a Location Manager
+func NewManager(cfg config.Configuration) (Manager, error) {
```
Comment (loicalbertin) [inline locations/location_mgr.go]: It is was not clear for me at the first glance that the manager may be cached.
Maybe if the function was named `GetManager` I may have a doubt and checked it.

File: commands/bootstrap/setup.go [hunk 5]
```diff
@@ -296,9 +296,9 @@ func waitForYorcServerUP(timeout time.Duration) error {
 	for {
 		response, err := client.Do(request)
 		if err == nil {
+			defer response.Body.Close()
```
Comment (loicalbertin) [inline commands/bootstrap/setup.go]: Is `defer` useful here?

File: locations/location_mgr.go [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+		locationMgr.hpAdapter = adapter.NewHostsPoolLocationAdapter(client)
+	}
+
+	return locationMgr, nil
+}
+
+// NewManagerWithSSHFactory creates a Location Manager with a given ssh factory
+//
+// Currently this is used for testing purpose to mock the ssh connection.
+func NewManagerWithSSHFactory(cfg config.Configuration, sshClientFactory hostspool.SSHClientFactory) (Manager, error) {
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply locations/location_mgr.go]: It makes sense !

File: commands/bootstrap/setup.go (hunk 5 above)
Comment (stebenoist) [reply commands/bootstrap/setup.go]: Fixed

**Pull request merged**